myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/OfferLifecycleManager.java

```diff
@@ -0,0 +1,86 @@
+package com.ebay.myriad.scheduler;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.inject.Inject;
+
+import org.apache.mesos.Protos;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Manages the Mesos offers tracked by Myriad.
+ */
+public class OfferLifecycleManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(
+      OfferLifecycleManager.class);
+
+  private Map<String, OfferFeed> offerFeedMap;
+
+  /**
+   * !!! Not thread safe !!!
+   */
+  private final Map<String, ConsumedOffer> consumedOfferMap;
+
+  private final NodeStore nodeStore;
+  private final MyriadDriver myriadDriver;
+
+  @Inject
+  public OfferLifecycleManager(NodeStore nodeStore,
+      MyriadDriver myriadDriver) {
+
+    this.offerFeedMap = new ConcurrentHashMap<>(200, 0.75f, 50);
+    this.consumedOfferMap = new HashMap<>(200, 0.75f);
+    this.nodeStore = nodeStore;
+    this.myriadDriver = myriadDriver;
+  }
+
+  public OfferFeed getOfferFeed(String hostname) {
+    return offerFeedMap.get(hostname);
+  }
+
+  public void declineOffer(Protos.Offer offer) {
+    myriadDriver.getDriver().declineOffer(offer.getId());
+    LOGGER.debug("Declined offer {}", offer.getId());
+  }
+
+  public void addOffers(List<Protos.Offer> offers) {
+    for (Protos.Offer offer : offers) {
+      String hostname = offer.getHostname();
+      Node node = nodeStore.getNode(hostname);
+      if (node != null) {
+        OfferFeed feed = offerFeedMap.get(hostname);
+        if (feed == null) {
+          feed = new OfferFeed();
+          offerFeedMap.put(hostname, feed);
+        }
+        feed.add(offer);
+
+        node.setSlaveId(offer.getSlaveId());
+
+        LOGGER.debug("addResourceOffers: caching offer for host {}, offer id {}",
+            hostname, offer.getId().getValue());
+      } else {
+        myriadDriver.getDriver().declineOffer(offer.getId());
+        LOGGER.debug("Declined offer for unregistered host {}", hostname);
+      }
+    }
```

Comment:
>OTOH, if the scheduling didn't result in new containers to be allocated, then Myriad simply >rejects the offer so that mesos can claim back the resources and offer them to other >frameworks.
Can you point me to were this happens?  It doesn't appear to be explicit I only see declineOffer called twice, but maybe I'm missing how it's happening.